### PR TITLE
PYMT-1657 Search Adjustments

### PR DESCRIPTION
**Pull request by merk**
This is a PR that contains half a dozen tickets worth of adjustments to achieve a new search approach.

Missing pieces: batching, pre-loading the batch entities rather than 1:1 querying, only reindex when mappings change

The "end result" of this pr can be seen in the README.md changes which describe how the SearchHandlers need to change.

### CODE REVIEW

File: src/Helpers/RegisteredSearchHandler.php [hunk 1]
```diff
@@ -31,6 +39,51 @@ public function getAll(): array
...
+
+        return $groupedSubscriptions;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getTransformableHandlerByKey(string $key): TransformableSearchHandlerInterface
+    {
+        if ($this->handlersByKey === null) {
```
Comment (rashmitsingh) [inline src/Helpers/RegisteredSearchHandler.php:71]: will this (stateful class) be a problem that helper is bound as a singleton?

File: src/Workers/EntityUpdateWorker.php [hunk 2]
```diff
@@ -3,56 +3,158 @@
...
+     */
+    private function shouldNotify(ChangeSubscription $subscription, ChangedEntity $change): bool
+    {
+        if ($change instanceof UpdatedEntity === true) {
+            $intersection = \array_intersect($subscription->getProperties(), $change->getChangedProperties());
+
+            return \count($intersection) > 0;
         }
+
+        return true;
```
Comment (rashmitsingh) [inline src/Workers/EntityUpdateWorker.php:164]: bases on the documentation on the method, should this not default to false?

File: tests/TestCase.php [hunk 3]
```diff
@@ -19,7 +19,7 @@
  *
  * @coversNothing
  */
-abstract class TestCase extends BaseTestCase
+abstract class TestCase extends UnitTestCase
```
Comment (rashmitsingh) [inline tests/TestCase.php:22]: 💯

**Review by rashmitsingh — COMMENTED** [review]
a couple of questions.

File: src/Helpers/RegisteredSearchHandler.php (hunk 1 above)
Comment (merk) [reply src/Helpers/RegisteredSearchHandler.php:71]: The only state we're storing here is based on the DI injections and is just a micro optimisation so we dont have to iterate every handler every time

File: src/Workers/EntityUpdateWorker.php (hunk 2 above)
Comment (merk) [reply src/Workers/EntityUpdateWorker.php:164]: If we didnt get an updated entity (we got a deletion instead) we dont have an array of changed properties to compare against - so we just have to notify every time

File: src/Workers/EntityUpdateWorker.php [hunk 4]
```diff
@@ -3,56 +3,164 @@
...
+     */
+    private function shouldNotify(ChangeSubscription $subscription, ChangedEntity $change): bool
+    {
+        // If the properties property on the subscription is null, we always notify
+        // of a change.
+        if ($subscription->getProperties() === null) {
+            return true;
         }
+
+        if ($change instanceof UpdatedEntity === true) {
```
Comment (brendonofficial) [inline src/Workers/EntityUpdateWorker.php:158]: missing the big `(` and `)`

**Review by brendonofficial — CHANGES_REQUESTED** [review]
Nice changes 👌

**Review by rashmitsingh — APPROVED** [review]
looks right.

**Review by brendonofficial — APPROVED** [review]

**Pull request merged**